criu/clone-noasan.c

```diff
@@ -27,14 +29,80 @@
  *	clone_noasan
  *	... wait for process to finish ...
  *	unlock_last_pid
+ *
+ * This locking is not required if clone3() is available. lock_last_pid()
+ * and unlock_last_pid() will be noops if clone3() is available.
  */
-int clone_noasan(int (*fn)(void *), int flags, void *arg)
+int clone_noasan(int (*fn)(void *), int flags, void *arg, pid_t pid)
 {
 	void *stack_ptr = (void *)round_down((unsigned long)&stack_ptr - 1024, 16);
+	struct _clone_args c_args = {};
+	long ret;
+
 	BUG_ON((flags & CLONE_VM) && !(flags & CLONE_VFORK));
 	/*
 	 * Reserve some bytes for clone() internal needs
 	 * and use as stack the address above this area.
 	 */
-	return clone(fn, stack_ptr, flags, arg);
+	if (pid == 0)
+		return clone(fn, stack_ptr, flags, arg);
+
+	pr_debug("Creating process using clone3()\n");
+
+	/*
+	 * The simple stack as configured for clone() above those not work
+	 * with clone3(). At the writing of this code it was not clear
+	 * where the difference is. One difference between clone() and
+	 * clone3() is that clone3() wants the stack address and the stack
+	 * size. The kernel does (stack + stack_size) and uses that value
+	 * as the stack pointer for the newly created process.
+	 */
+
+	/*
+	 * The stack_size seems to be pretty big and in the mmap() it is
+	 * even twice the size. By telling the kernel the stack is
+	 * STACK_SIZE and mmap()ing it twice the size the restore actually
+	 * works again. Any different combination did not work. This
+	 * is not understood.
+	 */
+
+#define STACK_SIZE 1024 * 1024 * 32
+
+	stack_ptr = mmap(NULL, STACK_SIZE * 2, PROT_READ | PROT_WRITE,
```

Comment:
we don't need to allocate a stack here, we can use the current one by the same way how it works for libc clone.

Reply:
Actually, when we are forking a new process with a specified pid, clone_flags can't contain CLONE_VM, this means that you can add clone_with_pid_noasan:

```
int clone_with_pid_noasan(int (*fn)(void *), int flags, void *arg, pid_t pid) {
  struct _clone_args c_args = {};
  pid_t pid;

  BUG_ON(flags & CLONE_VM);

  c_args.flags = flags;
  c_args.set_tid = ptr_to_u64(&pid);
  c_args.set_tid_size = 1;
  pid = syscall(__NR_clone3, &c_args, sizeof(c_args);
  if (pid == 0) {
    exit(fn(args));
  }
  return pid;
}
```

Reply:
Unfortunately the restored process crashes if using the stack the same way as with the libc clone().

Reply:
The restored process also crashes if using your clone_with_pid_noasan() example instead of my own assembler wrapper. Your code looks so much simpler than my assembler wrapper, but it does not work.

The interesting thing about the crash is, that not the process that is created with your code crashes, but the second thread of my test process segfaults when the first is restored with your code.

If I change `c_args.stack_size` in the thread restorer (pie/restorer.c) to `RESTORE_STACK_SIZE / 2` the restore works most of the time, but sometimes the restored thread segfaults.

Reply:
Can you show that code you're testing this with, please?

Reply:
@adrianreber I think something wrong in RUN_CLONE3_RESTORE_FN. I don't try to find what wrong in your code, I took RUN_CLONE_RESTORE_FN and modified it to use clone3. Here is my version and it works on my laptop:
https://github.com/avagin/criu/commit/2c9dc9d93ba3dc5a500138f1a3e3783269779fcb

Could you take a look at it?

Reply:
@avagin You are not setting the stack_size anywhere in your code. The kernel does:
```
        if (kargs->stack == 0) {
                if (kargs->stack_size > 0)
                        return false;
        } else {
                if (kargs->stack_size == 0)
                        return false;

                if (!access_ok((void __user *)kargs->stack, kargs->stack_size))
                        return false;

#if !defined(CONFIG_STACK_GROWSUP) && !defined(CONFIG_IA64)
                kargs->stack += kargs->stack_size;
#endif
        }
```
You should not be able to create a process using clone3() as you are setting stack to newsp but stack_size is 0.

Reply:
Depends on what kernel @avagin is on, but yes.